subprojects/gst-plugins-good/gst/rtp/gstrtph266pay.c

```diff
@@ -0,0 +1,629 @@
+ * Boston, MA 02110-1301, USA.
+ */
+
+/* references:
+ * - RTP Payload Format for Versatile Video Coding (VVC)
+ *   https://www.ietf.org/rfc/rfc9328.txt
+ * - RTP: A Transport Protocol for Real-Time Applications
+ *   https://www.ietf.org/rfc/rfc3550.txt
+ * - H.266: Versatile video coding
+ *   https://www.itu.int/rec/T-REC-H.266-202309-I
+ */
+
+#include "gstrtph266pay.h"
+#include <gst/rtp/gstrtppayloads.h>
+#include <gst/rtp/gstrtpelements.h>
+#include <gst/rtp/gstrtpbuffer.h>
+#include <gst/rtp/gstrtputils.h>
+#include <gst/base/gstadapter.h>
+
+#define GST_CAT_DEFAULT rtph266pay_debug
+GST_DEBUG_CATEGORY_STATIC (GST_CAT_DEFAULT);
+
+typedef enum
+{
+  NALU_TYPE_VPS = 14,
+  NALU_TYPE_SPS = 15,
+  NALU_TYPE_PPS = 16,
+  NALU_TYPE_PAPS = 17,
+  NALU_TYPE_SAPS = 18,
+  NALU_TYPE_AUD = 20,
+} NaluType;
+
+#define NALU_SC_MSK 0xffffff00
+#define NALU_SC_VAL 0x00000100
+#define NALU_SC_LEN 3
+#define NALU_HDR_LEN 2
+#define NALU_INVALID_XPS 0xFF
+#define NALU_IS_PARAMETER_SET(nalu) \
+  (((nalu)->type >= NALU_TYPE_VPS) && ((nalu)->type <= NALU_TYPE_SAPS))
+
+#define FU_TYPE 29
+#define FU_HDR_LEN (NALU_HDR_LEN + 1)   // PayloadHdr + FU header
+
+typedef struct
+{
+  GstBuffer *nalu_buf;
+  GstBuffer *hdr_buf;
+  GstBuffer *rbsp_buf;
+  guint16 size;                 // Size of the NALU (inluding its header)
+  NaluType type;                // nal_unit_type
+  guint8 xps_id;                // {vps_video,sps_seq,pps_pic,aps_adaptation}_parameter_set_id
+  gboolean au_start;
+  gboolean au_end;
+} Nalu;
+
+#define NALU_PTR_FORMAT \
+  "p, size: %u, type: %u, xps_id: %u, au_start: %s, au_end: %s"
+#define NALU_ARGS(nalu) \
+  (nalu), (nalu)->size, (nalu)->type, (nalu)->xps_id, (nalu)->au_start ? "true" : "false", (nalu)->au_end ? "true" : "false"
+
+typedef enum
+{
+  ALIGNMENT_AU,
+  ALIGNMENT_NAL,
+  ALIGNMENT_UNKOWN,
+} Alignment;
+
+struct _GstRtpH266Pay
+{
+  GstRTPBasePayload payload;
+
+  GstAdapter *adapter;
+  GQueue nalus;
+  Alignment alignment;
+};
+
+#define gst_rtp_h266_pay_parent_class parent_class
+G_DEFINE_TYPE (GstRtpH266Pay, gst_rtp_h266_pay, GST_TYPE_RTP_BASE_PAYLOAD);
+GST_ELEMENT_REGISTER_DEFINE_WITH_CODE (rtph266pay, "rtph266pay",
+    GST_RANK_SECONDARY, GST_TYPE_RTP_H266_PAY, rtp_element_init (plugin));
+
+static GstStaticPadTemplate sink_template =
+GST_STATIC_PAD_TEMPLATE ("sink", GST_PAD_SINK, GST_PAD_ALWAYS,
+    GST_STATIC_CAPS ("video/x-h266, stream-format = (string) byte-stream, "
+        "alignment = (string) { nal, au }"));
+
+static GstStaticPadTemplate src_template = GST_STATIC_PAD_TEMPLATE ("src",
+    GST_PAD_SRC,
+    GST_PAD_ALWAYS,
+    GST_STATIC_CAPS ("application/x-rtp, "
+        "media = (string) \"video\", "
+        "payload = (int) " GST_RTP_PAYLOAD_DYNAMIC_STRING ", "
+        "clock-rate = (int) 90000, " "encoding-name = (string) \"H266\"")
+    );
+
+enum
+{
+  PROP_0,
+};
+
+// GObject methods
+static void _finalize (GObject * object);
```

Comment:
What do you think about ordering functions from the child "namespace" to parent? This way we can get rid of (most of) these function declaration.

The order would be as follows:
1. Nalu functions
2. GstRtpH266Pay functions
3. GstRTPBasePayload functions
4. GstElement functions
5. GObject functions, i.e _set_property, _get_property
6. GstRtpH266Pay initialization functions: gst_rtp_h266_pay_class_init and gst_rtp_h266_pay_init

Reply:
I think having these function declarations is something good. It provides a quick overview of what you're going to find in this file, and also allows to have the definitions in any order.

Why do you want to get rid of them?

Reply:
Not deleting itself, but just because you put the functions in this order from child to parent, you know that GObjects are at the very bottom, and child classes are at the top. The consequence of this is that function declaration on the top are removed.